src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -355,6 +360,16 @@ public static bool IsUnnecessaryCast(this CastExpressionSyntax cast, SemanticMod
                 // Explicit reference conversions can cause an exception or data loss, hence can never be removed.
                 return false;
             }
+            else if (expressionToCastType.IsExplicit && expressionToCastType.IsNumeric && IsInUncheckedContext(cast))
+            {
+                // Don't remove any explicit numeric casts in unchecked context.
+                return false;
+            }
+            else if (expressionToCastType.IsPointer)
+            {
+                // Don't remove any non-identity pointer conversions.
+                return expressionType != null && expressionType.Equals(outerType);
```

Comment:
Hmm, I have taken the most conservative approach here :) Let me open a new issue to properly deal with cast simplification involving pointer conversions. Probably you and/or other unsafe code experts can guide/review the implementation.